Panic causes "thread panicked at 'internal error: entered unreachable code'"

Hi, thank you for this very useful library! However, I've found a way to enter a supposedly unreachable state.

Here's the relevant part of the backtrace:

```
thread 'main' panicked at 'internal error: entered unreachable code', /home/user/.cargo/git/checkouts/moka-6ea430727379b61e/19b6925/src/future/value_initializer.rs:101:29
stack backtrace:
   0: rust_begin_unwind
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/std/src/panicking.rs:515:5
   1: core::panicking::panic_fmt
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/panicking.rs:92:14
   2: core::panicking::panic
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/panicking.rs:50:5
   3: moka::future::value_initializer::ValueInitializer<K,V,S>::try_init_or_read::{{closure}}
             at /home/user/.cargo/git/checkouts/moka-6ea430727379b61e/19b6925/src/future/value_initializer.rs:101:29
   4: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/future/mod.rs:80:19
   5: moka::future::cache::Cache<K,V,S>::get_or_try_insert_with_hash_and_fun::{{closure}}
             at /home/user/.cargo/git/checkouts/moka-6ea430727379b61e/19b6925/src/future/cache.rs:640:15
   6: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/future/mod.rs:80:19
   7: moka::future::cache::Cache<K,V,S>::get_or_try_insert_with::{{closure}}
             at /home/user/.cargo/git/checkouts/moka-6ea430727379b61e/19b6925/src/future/cache.rs:445:9
```

This issue seems to occur when two threads attempt to fetch the same key and the thread on which the value initialization was started panics. I've built a simple program to demonstrate this issue:

```rust
use tokio;
use std::sync::Arc;

#[tokio::main]
async fn main() {
    let cache: Arc<moka::future::Cache<i32, i32>> = Arc::new(moka::future::Cache::new(16));
    let semaphore = Arc::new(tokio::sync::Semaphore::new(0));
    {
        let cache_ref = cache.clone();
        let semaphore_ref = semaphore.clone();
        tokio::task::spawn(async move {
            cache_ref.get_or_try_insert_with::<_, ()>(1, async move {
                semaphore_ref.add_permits(1);
                tokio::time::sleep(tokio::time::Duration::from_millis(1000));
                panic!("Panic during get_or_try_insert_with");
                Ok(10)
            }).await;
        });
    }
    semaphore.acquire().await;
    cache.get_or_try_insert_with::<_, ()>(1, async move {
        println!("Async block in second get_or_try_insert_with called");
        Ok(5)
    }).await;
}
```

I think the unreachable code is being hit because the write lock obtained at [line 71 of value_initializer.rs](https://github.com/moka-rs/moka/blob/19b6925d9e854374f97adecb074ddc192c47ee6b/src/sync/value_initializer.rs#L71) gets dropped without any value being written ([here](https://github.com/moka-rs/moka/blob/19b6925d9e854374f97adecb074ddc192c47ee6b/src/sync/value_initializer.rs#L78) or [here](https://github.com/moka-rs/moka/blob/19b6925d9e854374f97adecb074ddc192c47ee6b/src/sync/value_initializer.rs#L83)).

I see two possible options for handling this issue:
1. `unreachable()` can be replaced with a panic with a more helpful message.
2. Perhaps the existing waiter should be removed and a new waiter should be created and started using the future passed in the second call?

> I think we can take this option by using the catch_unwind method from the futures crate. 

Unless I am missing something, using `catch_unwind` would impose a lot of restrictions of what can happen inside the future. For instance, if I modify my example to use `catch_unwind`, it won't compile because of internal mutability in the `Semaphore`:

```rust
use futures::future::FutureExt;
use tokio;
use std::sync::Arc;
use std::any::Any;
use futures::TryStreamExt;
use tokio::sync::Mutex;
use futures::TryFutureExt;

#[tokio::main]
async fn main() {
    let cache: Arc<moka::future::Cache<i32, i32>> = Arc::new(moka::future::Cache::new(16));
    let semaphore = Arc::new(tokio::sync::Semaphore::new(0));
    {
        let cache_ref = cache.clone();
        let semaphore_ref = semaphore.clone();
        tokio::task::spawn(async move {
            cache_ref.get_or_try_insert_with::<_, tokio::sync::Mutex<Box<(dyn Any + std::marker::Send + 'static)>>>(1, async move {
                semaphore_ref.add_permits(1);
                tokio::time::sleep(tokio::time::Duration::from_millis(1000));
                panic!("Panic during get_or_try_insert_with");
                10
            }.catch_unwind().map_err(|pn| Mutex::new(pn))).await;
        });
    }
    semaphore.acquire().await;
    cache.get_or_try_insert_with::<_, ()>(1, async move {
        println!("Async block in second get_or_try_insert_with called");
        Ok(5)
    }).await;
}
```

And here's the compiler error:

```
error[E0277]: the type `UnsafeCell<tokio::sync::batch_semaphore::Waitlist>` may contain interior mutability and a reference may not be safely transferrable across a catch_unwind boundary
  --> src/main.rs:22:15
   |
22 |             }.catch_unwind().map_err(|pn| Mutex::new(pn))).await;
   |               ^^^^^^^^^^^^ `UnsafeCell<tokio::sync::batch_semaphore::Waitlist>` may contain interior mutability and a reference may not be safely transferrable across a catch_unwind boundary
   |
   = help: within `Semaphore`, the trait `RefUnwindSafe` is not implemented for `UnsafeCell<tokio::sync::batch_semaphore::Waitlist>`
note: captured value does not implement `UnwindSafe`
  --> src/main.rs:18:17
   |
18 |                 semaphore_ref.add_permits(1);
   |                 ^^^^^^^^^^^^^ has type `Arc<Semaphore>` which does not implement `UnwindSafe`

```

However, it should be possible to detect the panic  on the other thread  without directly catching it by handling the case currently marked as unreachable in `try_init_or_read`, since `None` should only occur in cases when the initialization future has terminated before writing anything. A hacky solution would be to remove the waiter and call `try_init_or_read` recursively, but that could result in stack overflows.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,7 +7,7 @@
 - Fix a bug in `get_or_insert_with` and `get_or_try_insert_with` methods of
   `future::Cache` and `sync::Cache`; a panic in the `init` future/closure
   causes subsequent calls on the same key to get "unreachable code" panics.
-  ([#41][gh-issue-0043])
+  ([#43][gh-issue-0043])
 
 ### Changed
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -2,6 +2,13 @@
 
 ## Version 0.6.0 (Unreleased)
 
+### Fixed
+
+- Fix a bug in `get_or_insert_with` and `get_or_try_insert_with` methods of
+  `future::Cache` and `sync::Cache`; a panic in the `init` future/closure
+  causes subsequent calls on the same key to get "unreachable code" panics.
+  ([#41][gh-issue-0043])
+
 ### Changed
 
 - Change `get_or_try_insert_with` to return a concrete error type rather
@@ -133,6 +140,7 @@
 
 [resolving-error-on-32bit]: https://github.com/moka-rs/moka#resolving-compile-errors-on-some-32-bit-platforms
 
+[gh-issue-0043]: https://github.com/moka-rs/moka/issues/43/
 [gh-pull-0042]: https://github.com/moka-rs/moka/pull/42/
 [gh-issue-0038]: https://github.com/moka-rs/moka/issues/38/
 [gh-pull-0037]: https://github.com/moka-rs/moka/pull/37/

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,6 +1,6 @@
 # Moka &mdash; Change Log
 
-## Version 0.6.0 (Unreleased)
+## Version 0.6.0
 
 ### Fixed
 
@@ -13,6 +13,7 @@
 
 - Change `get_or_try_insert_with` to return a concrete error type rather
   than a trait object. ([#23][gh-pull-0023], [#37][gh-pull-0037])
+
 
 ## Version 0.5.4
 

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -54,7 +54,7 @@ async-std = { version = "1", default-features = false, features = ["attributes"]
 getrandom = "0.2"
 reqwest = "0.11"
 skeptic = "0.13"
-tokio = { version = "1", features = ["rt-multi-thread", "macros" ] }
+tokio = { version = "1", features = ["rt-multi-thread", "macros", "sync", "time" ] }
 
 [target.'cfg(trybuild)'.dev-dependencies]
 trybuild = "1.0"

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -23,7 +23,7 @@ features = ["future"]
 default = ["atomic64"]
 
 # Enable this feature to use `moka::future::Cache`.
-future = ["async-io", "async-lock"]
+future = ["async-io", "async-lock", "futures"]
 
 # This feature is enabled by default. Disable it when the target platform does not
 # support `std::sync::atomic::AtomicU64`. (e.g. `armv5te-unknown-linux-musleabi`
@@ -45,12 +45,12 @@ uuid = { version = "0.8", features = ["v4"] }
 # Optional dependencies
 async-io = { version = "1.4", optional = true }
 async-lock = { version = "2.4", optional = true }
+futures = { version = "0.3", optional = true }
 
 [dev-dependencies]
 actix-rt2 = { package = "actix-rt", version = "2", default-features = false }
 actix-rt1 = { package = "actix-rt", version = "1", default-features = false }
 async-std = { version = "1", default-features = false, features = ["attributes"] }
-futures = "0.3"
 getrandom = "0.2"
 reqwest = "0.11"
 skeptic = "0.13"

--- a/src/future/cache.rs
+++ b/src/future/cache.rs
@@ -1339,7 +1339,7 @@ mod tests {
             });
         }
         let _ = semaphore.acquire().await.expect("semaphore acquire failed");
-        assert_eq!(cache.get_or_insert_with(1, async move { 5 }).await, 5);
+        assert_eq!(cache.get_or_insert_with(1, async { 5 }).await, 5);
     }
 
     #[tokio::test]
@@ -1364,8 +1364,7 @@ mod tests {
         }
         let _ = semaphore.acquire().await.expect("semaphore acquire failed");
         assert_eq!(
-            cache.get_or_try_insert_with(1, async move { Ok(5) }).await
-                as Result<_, Arc<Infallible>>,
+            cache.get_or_try_insert_with(1, async { Ok(5) }).await as Result<_, Arc<Infallible>>,
             Ok(5)
         );
     }

--- a/src/future/cache.rs
+++ b/src/future/cache.rs
@@ -329,7 +329,7 @@ where
     ///
     /// **A Sample Result**
     ///
-    /// - The async black resolved exactly once by task 3.
+    /// - The `init` future (async black) was resolved exactly once by task 3.
     /// - Other tasks were blocked until task 3 inserted the value.
     ///
     /// ```console
@@ -343,6 +343,14 @@ where
     /// Task 1 got the value. (len: 10485760)
     /// Task 2 got the value. (len: 10485760)
     /// ```
+    ///
+    /// # Panics
+    ///
+    /// This method panics when the `init` future has been panicked. When it happens,
+    /// only the caller whose `init` future panicked will get the panic (e.g. only
+    /// task 3 in the above sample). If there are other calls in progress (e.g. task
+    /// 0, 1 and 2 above), this method will restart and resolve one of the remaining
+    /// `init` futures.
     ///
     pub async fn get_or_insert_with<F>(&self, key: K, init: F) -> V
     where
@@ -420,7 +428,7 @@ where
     ///
     /// **A Sample Result**
     ///
-    /// - `get_html()` called exactly once by task 2.
+    /// - `get_html()` was called exactly once by task 2.
     /// - Other tasks were blocked until task 2 inserted the value.
     ///
     /// ```console
@@ -434,6 +442,14 @@ where
     /// Task 0 got the value. (len: 19419)
     /// Task 3 got the value. (len: 19419)
     /// ```
+    ///
+    /// # Panics
+    ///
+    /// This method panics when the `init` future has been panicked. When it happens,
+    /// only the caller whose `init` future panicked will get the panic (e.g. only
+    /// task 2 in the above sample). If there are other calls in progress (e.g. task
+    /// 0, 1 and 3 above), this method will restart and resolve one of the remaining
+    /// `init` futures.
     ///
     pub async fn get_or_try_insert_with<F, E>(&self, key: K, init: F) -> Result<V, Arc<E>>
     where
@@ -750,7 +766,7 @@ mod tests {
     use crate::{common::time::Clock, future::CacheBuilder};
 
     use async_io::Timer;
-    use std::time::Duration;
+    use std::{convert::Infallible, sync::Arc, time::Duration};
 
     #[tokio::test]
     async fn basic_single_async_task() {
@@ -1300,5 +1316,57 @@ mod tests {
         };
 
         futures::join!(task1, task2, task3, task4, task5, task6, task7, task8);
+    }
+
+    #[tokio::test]
+    // https://github.com/moka-rs/moka/issues/43
+    async fn handle_panic_in_get_or_insert_with() {
+        use tokio::time::{sleep, Duration};
+
+        let cache = Cache::new(16);
+        let semaphore = Arc::new(tokio::sync::Semaphore::new(0));
+        {
+            let cache_ref = cache.clone();
+            let semaphore_ref = semaphore.clone();
+            tokio::task::spawn(async move {
+                let _ = cache_ref
+                    .get_or_insert_with(1, async move {
+                        semaphore_ref.add_permits(1);
+                        sleep(Duration::from_millis(50)).await;
+                        panic!("Panic during get_or_try_insert_with");
+                    })
+                    .await;
+            });
+        }
+        let _ = semaphore.acquire().await.expect("semaphore acquire failed");
+        assert_eq!(cache.get_or_insert_with(1, async move { 5 }).await, 5);
+    }
+
+    #[tokio::test]
+    // https://github.com/moka-rs/moka/issues/43
+    async fn handle_panic_in_get_or_try_insert_with() {
+        use tokio::time::{sleep, Duration};
+
+        let cache = Cache::new(16);
+        let semaphore = Arc::new(tokio::sync::Semaphore::new(0));
+        {
+            let cache_ref = cache.clone();
+            let semaphore_ref = semaphore.clone();
+            tokio::task::spawn(async move {
+                let _ = cache_ref
+                    .get_or_try_insert_with(1, async move {
+                        semaphore_ref.add_permits(1);
+                        sleep(Duration::from_millis(50)).await;
+                        panic!("Panic during get_or_try_insert_with");
+                    })
+                    .await as Result<_, Arc<Infallible>>;
+            });
+        }
+        let _ = semaphore.acquire().await.expect("semaphore acquire failed");
+        assert_eq!(
+            cache.get_or_try_insert_with(1, async move { Ok(5) }).await
+                as Result<_, Arc<Infallible>>,
+            Ok(5)
+        );
     }
 }

--- a/src/future/value_initializer.rs
+++ b/src/future/value_initializer.rs
@@ -82,7 +82,7 @@ where
 
     /// # Panics
     /// Panics if the `init` future has been panicked.
-    async fn do_try_init<'a, F, FO, C, E>(
+    async fn do_try_init<'a, F, O, C, E>(
         &self,
         key: &'a Arc<K>,
         type_id: TypeId,
@@ -90,8 +90,8 @@ where
         mut post_init: C,
     ) -> InitResult<V, E>
     where
-        F: Future<Output = FO>,
-        C: FnMut(&'a Arc<K>, FO, &mut WaiterValue<V>) -> InitResult<V, E>,
+        F: Future<Output = O>,
+        C: FnMut(&'a Arc<K>, O, &mut WaiterValue<V>) -> InitResult<V, E>,
         E: Send + Sync + 'static,
     {
         use futures::future::FutureExt;
@@ -107,15 +107,15 @@ where
                     // Our waiter was inserted. Let's resolve the init future.
                     // Catching panic is safe here as we do not try to resolve the future again.
                     match AssertUnwindSafe(init).catch_unwind().await {
-                        // resolved.
+                        // Resolved.
                         Ok(value) => return post_init(key, value, &mut lock),
-                        // panicked.
+                        // Panicked.
                         Err(payload) => {
                             *lock = None;
                             // Remove the waiter so that others can retry.
                             self.remove_waiter(key, type_id);
                             resume_unwind(payload);
-                        }
+                        } // The lock will be unlock here.
                     }
                 }
                 Some(res) => {

--- a/src/future/value_initializer.rs
+++ b/src/future/value_initializer.rs
@@ -98,6 +98,9 @@ where
         use std::panic::{resume_unwind, AssertUnwindSafe};
         use InitResult::*;
 
+        const MAX_RETRIES: usize = 200;
+        let mut retries = 0;
+
         loop {
             let waiter = Arc::new(RwLock::new(None));
             let mut lock = waiter.write().await;
@@ -115,7 +118,7 @@ where
                             // Remove the waiter so that others can retry.
                             self.remove_waiter(key, type_id);
                             resume_unwind(payload);
-                        } // The lock will be unlock here.
+                        } // The lock will be unlocked here.
                     }
                 }
                 Some(res) => {
@@ -126,8 +129,19 @@ where
                         Some(Ok(value)) => return ReadExisting(value.clone()),
                         Some(Err(e)) => return InitErr(Arc::clone(e).downcast().unwrap()),
                         // None means somebody else's init future has been panicked.
-                        // Retry from the beginning.
-                        None => continue,
+                        None => {
+                            retries += 1;
+                            if retries < MAX_RETRIES {
+                                // Retry from the beginning.
+                                continue;
+                            } else {
+                                panic!(
+                                    r#"Too many retries. Tried to read the return value from the `init` \
+                                future but failed {} times. Maybe the `init` kept panicking?"#,
+                                    retries
+                                );
+                            }
+                        }
                     }
                 }
             }

--- a/src/future/value_initializer.rs
+++ b/src/future/value_initializer.rs
@@ -7,7 +7,8 @@ use std::{
 };
 
 type ErrorObject = Arc<dyn Any + Send + Sync + 'static>;
-type Waiter<V> = Arc<RwLock<Option<Result<V, ErrorObject>>>>;
+type WaiterValue<V> = Option<Result<V, ErrorObject>>;
+type Waiter<V> = Arc<RwLock<WaiterValue<V>>>;
 
 pub(crate) enum InitResult<V, E> {
     Initialized(V),
@@ -35,70 +36,99 @@ where
         }
     }
 
+    /// # Panics
+    /// Panics if the `init` future has been panicked.
     pub(crate) async fn init_or_read<F>(&self, key: Arc<K>, init: F) -> InitResult<V, ()>
     where
         F: Future<Output = V>,
     {
-        use InitResult::*;
+        // This closure will be called after the init closure has returned a value.
+        // It will convert the returned value (from init) into an InitResult.
+        let post_init = |_key, value: V, lock: &mut WaiterValue<V>| {
+            *lock = Some(Ok(value.clone()));
+            InitResult::Initialized(value)
+        };
 
         let type_id = TypeId::of::<()>();
-        let waiter = Arc::new(RwLock::new(None));
-        let mut lock = waiter.write().await;
-
-        match self.try_insert_waiter(&key, type_id, &waiter) {
-            None => {
-                // Our waiter was inserted. Let's resolve the init future.
-                let value = init.await;
-                *lock = Some(Ok(value.clone()));
-                Initialized(value)
-            }
-            Some(res) => {
-                // Somebody else's waiter already exists. Drop our write lock and wait
-                // for a read lock to become available.
-                std::mem::drop(lock);
-                match &*res.read().await {
-                    Some(Ok(value)) => ReadExisting(value.clone()),
-                    Some(Err(_)) | None => unreachable!(),
-                }
-            }
-        }
+        self.do_try_init(&key, type_id, init, post_init).await
     }
 
+    /// # Panics
+    /// Panics if the `init` future has been panicked.
     pub(crate) async fn try_init_or_read<F, E>(&self, key: Arc<K>, init: F) -> InitResult<V, E>
     where
         F: Future<Output = Result<V, E>>,
         E: Send + Sync + 'static,
     {
+        let type_id = TypeId::of::<E>();
+
+        // This closure will be called after the init closure has returned a value.
+        // It will convert the returned value (from init) into an InitResult.
+        let post_init = |key, value: Result<V, E>, lock: &mut WaiterValue<V>| match value {
+            Ok(value) => {
+                *lock = Some(Ok(value.clone()));
+                InitResult::Initialized(value)
+            }
+            Err(e) => {
+                let err: ErrorObject = Arc::new(e);
+                *lock = Some(Err(Arc::clone(&err)));
+                self.remove_waiter(key, type_id);
+                InitResult::InitErr(err.downcast().unwrap())
+            }
+        };
+
+        self.do_try_init(&key, type_id, init, post_init).await
+    }
+
+    /// # Panics
+    /// Panics if the `init` future has been panicked.
+    async fn do_try_init<'a, F, FO, C, E>(
+        &self,
+        key: &'a Arc<K>,
+        type_id: TypeId,
+        init: F,
+        mut post_init: C,
+    ) -> InitResult<V, E>
+    where
+        F: Future<Output = FO>,
+        C: FnMut(&'a Arc<K>, FO, &mut WaiterValue<V>) -> InitResult<V, E>,
+        E: Send + Sync + 'static,
+    {
+        use futures::future::FutureExt;
+        use std::panic::{resume_unwind, AssertUnwindSafe};
         use InitResult::*;
 
-        let type_id = TypeId::of::<E>();
-        let waiter = Arc::new(RwLock::new(None));
-        let mut lock = waiter.write().await;
+        loop {
+            let waiter = Arc::new(RwLock::new(None));
+            let mut lock = waiter.write().await;
 
-        match self.try_insert_waiter(&key, type_id, &waiter) {
-            None => {
-                // Our waiter was inserted. Let's resolve the init future.
-                match init.await {
-                    Ok(value) => {
-                        *lock = Some(Ok(value.clone()));
-                        Initialized(value)
-                    }
-                    Err(e) => {
-                        let err: ErrorObject = Arc::new(e);
-                        *lock = Some(Err(Arc::clone(&err)));
-                        self.remove_waiter(&key, type_id);
-                        InitErr(err.downcast().unwrap())
+            match self.try_insert_waiter(key, type_id, &waiter) {
+                None => {
+                    // Our waiter was inserted. Let's resolve the init future.
+                    // Catching panic is safe here as we do not try to resolve the future again.
+                    match AssertUnwindSafe(init).catch_unwind().await {
+                        // resolved.
+                        Ok(value) => return post_init(key, value, &mut lock),
+                        // panicked.
+                        Err(payload) => {
+                            *lock = None;
+                            // Remove the waiter so that others can retry.
+                            self.remove_waiter(key, type_id);
+                            resume_unwind(payload);
+                        }
                     }
                 }
-            }
-            Some(res) => {
-                // Somebody else's waiter already exists. Drop our write lock and wait
-                // for a read lock to become available.
-                std::mem::drop(lock);
-                match &*res.read().await {
-                    Some(Ok(value)) => ReadExisting(value.clone()),
-                    Some(Err(e)) => InitErr(Arc::clone(e).downcast().unwrap()),
-                    None => unreachable!(),
+                Some(res) => {
+                    // Somebody else's waiter already exists. Drop our write lock and wait
+                    // for a read lock to become available.
+                    std::mem::drop(lock);
+                    match &*res.read().await {
+                        Some(Ok(value)) => return ReadExisting(value.clone()),
+                        Some(Err(e)) => return InitErr(Arc::clone(e).downcast().unwrap()),
+                        // None means somebody else's init future has been panicked.
+                        // Retry from the beginning.
+                        None => continue,
+                    }
                 }
             }
         }
@@ -110,6 +140,7 @@ where
         self.waiters.remove(&(key, type_id));
     }
 
+    #[inline]
     fn try_insert_waiter(
         &self,
         key: &Arc<K>,

--- a/src/sync/cache.rs
+++ b/src/sync/cache.rs
@@ -316,6 +316,14 @@ where
     /// Thread 3 got the value. (len: 10485760)
     /// ```
     ///
+    /// # Panics
+    ///
+    /// This method panics when the `init` closure has been panicked. When it
+    /// happens, only the caller whose `init` closure panicked will get the panic
+    /// (e.g. only thread 1 in the above sample). If there are other calls in
+    /// progress (e.g. thread 0, 2 and 3 above), this method will restart and resolve
+    /// one of the remaining `init` closure.
+    ///
     pub fn get_or_insert_with(&self, key: K, init: impl FnOnce() -> V) -> V {
         let hash = self.base.hash(&key);
         let key = Arc::new(key);
@@ -418,6 +426,14 @@ where
     /// Thread 1 got the value. (len: 1466)
     /// Thread 3 got the value. (len: 1466)
     /// ```
+    ///
+    /// # Panics
+    ///
+    /// This method panics when the `init` closure has been panicked. When it
+    /// happens, only the caller whose `init` closure panicked will get the panic
+    /// (e.g. only thread 1 in the above sample). If there are other calls in
+    /// progress (e.g. thread 0, 2 and 3 above), this method will restart and resolve
+    /// one of the remaining `init` closure.
     ///
     pub fn get_or_try_insert_with<F, E>(&self, key: K, init: F) -> Result<V, Arc<E>>
     where
@@ -648,7 +664,7 @@ mod tests {
     use super::{Cache, ConcurrentCacheExt};
     use crate::{common::time::Clock, sync::CacheBuilder};
 
-    use std::time::Duration;
+    use std::{convert::Infallible, sync::Arc, time::Duration};
 
     #[test]
     fn basic_single_thread() {
@@ -1137,5 +1153,54 @@ mod tests {
         ] {
             t.join().expect("Failed to join");
         }
+    }
+
+    #[test]
+    // https://github.com/moka-rs/moka/issues/43
+    fn handle_panic_in_get_or_insert_with() {
+        use std::{sync::Barrier, thread};
+
+        let cache = Cache::new(16);
+        let barrier = Arc::new(Barrier::new(2));
+        {
+            let cache_ref = cache.clone();
+            let barrier_ref = barrier.clone();
+            thread::spawn(move || {
+                let _ = cache_ref.get_or_insert_with(1, || {
+                    barrier_ref.wait();
+                    thread::sleep(Duration::from_millis(50));
+                    panic!("Panic during get_or_try_insert_with");
+                });
+            });
+        }
+
+        barrier.wait();
+        assert_eq!(cache.get_or_insert_with(1, || 5), 5);
+    }
+
+    #[test]
+    // https://github.com/moka-rs/moka/issues/43
+    fn handle_panic_in_get_or_try_insert_with() {
+        use std::{sync::Barrier, thread};
+
+        let cache = Cache::new(16);
+        let barrier = Arc::new(Barrier::new(2));
+        {
+            let cache_ref = cache.clone();
+            let barrier_ref = barrier.clone();
+            thread::spawn(move || {
+                let _ = cache_ref.get_or_try_insert_with(1, || {
+                    barrier_ref.wait();
+                    thread::sleep(Duration::from_millis(50));
+                    panic!("Panic during get_or_try_insert_with");
+                }) as Result<_, Arc<Infallible>>;
+            });
+        }
+
+        barrier.wait();
+        assert_eq!(
+            cache.get_or_try_insert_with(1, || Ok(5)) as Result<_, Arc<Infallible>>,
+            Ok(5)
+        );
     }
 }

--- a/src/sync/cache.rs
+++ b/src/sync/cache.rs
@@ -301,7 +301,7 @@ where
     ///
     /// **Result**
     ///
-    /// - The `init` closure called exactly once by thread 1.
+    /// - The `init` closure was called exactly once by thread 1.
     /// - Other threads were blocked until thread 1 inserted the value.
     ///
     /// ```console
@@ -412,7 +412,7 @@ where
     ///
     /// **Result**
     ///
-    /// - `get_file_size()` called exactly once by thread 1.
+    /// - `get_file_size()` was called exactly once by thread 1.
     /// - Other threads were blocked until thread 1 inserted the value.
     ///
     /// ```console

--- a/src/sync/value_initializer.rs
+++ b/src/sync/value_initializer.rs
@@ -93,6 +93,9 @@ where
         use std::panic::{catch_unwind, resume_unwind, AssertUnwindSafe};
         use InitResult::*;
 
+        const MAX_RETRIES: usize = 200;
+        let mut retries = 0;
+
         loop {
             let waiter = Arc::new(RwLock::new(None));
             let mut lock = waiter.write();
@@ -110,7 +113,7 @@ where
                             // Remove the waiter so that others can retry.
                             self.remove_waiter(key, type_id);
                             resume_unwind(payload);
-                        } // The lock will be unlock here.
+                        } // The write lock will be unlocked here.
                     }
                 }
                 Some(res) => {
@@ -120,9 +123,20 @@ where
                     match &*res.read() {
                         Some(Ok(value)) => return ReadExisting(value.clone()),
                         Some(Err(e)) => return InitErr(Arc::clone(e).downcast().unwrap()),
-                        // None means somebody else's init future has been panicked.
-                        // Retry from the beginning.
-                        None => continue,
+                        // None means somebody else's init closure has been panicked.
+                        None => {
+                            retries += 1;
+                            if retries < MAX_RETRIES {
+                                // Retry from the beginning.
+                                continue;
+                            } else {
+                                panic!(
+                                    r#"Too many retries. Tried to read the return value from the `init` \
+                                closure but failed {} times. Maybe the `init` kept panicking?"#,
+                                    retries
+                                );
+                            }
+                        }
                     }
                 }
             }

--- a/src/sync/value_initializer.rs
+++ b/src/sync/value_initializer.rs
@@ -6,7 +6,8 @@ use std::{
 };
 
 type ErrorObject = Arc<dyn Any + Send + Sync + 'static>;
-type Waiter<V> = Arc<RwLock<Option<Result<V, ErrorObject>>>>;
+type WaiterValue<V> = Option<Result<V, ErrorObject>>;
+type Waiter<V> = Arc<RwLock<WaiterValue<V>>>;
 
 pub(crate) enum InitResult<V, E> {
     Initialized(V),
@@ -34,66 +35,95 @@ where
         }
     }
 
+    /// # Panics
+    /// Panics if the `init` future has been panicked.
     pub(crate) fn init_or_read(&self, key: Arc<K>, init: impl FnOnce() -> V) -> InitResult<V, ()> {
-        use InitResult::*;
+        // This closure will be called after the init closure has returned a value.
+        // It will convert the returned value (from init) into an InitResult.
+        let post_init = |_key, value: V, lock: &mut WaiterValue<V>| {
+            *lock = Some(Ok(value.clone()));
+            InitResult::Initialized(value)
+        };
 
-        let waiter = Arc::new(RwLock::new(None));
-        let mut lock = waiter.write();
-
-        match self.try_insert_waiter(&key, TypeId::of::<()>(), &waiter) {
-            None => {
-                // Our waiter was inserted. Let's resolve the init closure.
-                let value = init();
-                *lock = Some(Ok(value.clone()));
-                Initialized(value)
-            }
-            Some(res) => {
-                // Somebody else's waiter already exists. Drop our write lock and wait
-                // for a read lock to become available.
-                std::mem::drop(lock);
-                match &*res.read() {
-                    Some(Ok(value)) => ReadExisting(value.clone()),
-                    Some(Err(_)) | None => unreachable!(),
-                }
-            }
-        }
+        let type_id = TypeId::of::<()>();
+        self.do_try_init(&key, type_id, init, post_init)
     }
 
+    /// # Panics
+    /// Panics if the `init` future has been panicked.
     pub(crate) fn try_init_or_read<F, E>(&self, key: Arc<K>, init: F) -> InitResult<V, E>
     where
         F: FnOnce() -> Result<V, E>,
         E: Send + Sync + 'static,
     {
+        let type_id = TypeId::of::<E>();
+
+        // This closure will be called after the init closure has returned a value.
+        // It will convert the returned value (from init) into an InitResult.
+        let post_init = |key, value: Result<V, E>, lock: &mut WaiterValue<V>| match value {
+            Ok(value) => {
+                *lock = Some(Ok(value.clone()));
+                InitResult::Initialized(value)
+            }
+            Err(e) => {
+                let err: ErrorObject = Arc::new(e);
+                *lock = Some(Err(Arc::clone(&err)));
+                self.remove_waiter(key, type_id);
+                InitResult::InitErr(err.downcast().unwrap())
+            }
+        };
+
+        self.do_try_init(&key, type_id, init, post_init)
+    }
+
+    /// # Panics
+    /// Panics if the `init` future has been panicked.
+    fn do_try_init<'a, F, O, C, E>(
+        &self,
+        key: &'a Arc<K>,
+        type_id: TypeId,
+        init: F,
+        mut post_init: C,
+    ) -> InitResult<V, E>
+    where
+        F: FnOnce() -> O,
+        C: FnMut(&'a Arc<K>, O, &mut WaiterValue<V>) -> InitResult<V, E>,
+        E: Send + Sync + 'static,
+    {
+        use std::panic::{catch_unwind, resume_unwind, AssertUnwindSafe};
         use InitResult::*;
 
-        let type_id = TypeId::of::<E>();
-        let waiter = Arc::new(RwLock::new(None));
-        let mut lock = waiter.write();
+        loop {
+            let waiter = Arc::new(RwLock::new(None));
+            let mut lock = waiter.write();
 
-        match self.try_insert_waiter(&key, type_id, &waiter) {
-            None => {
-                // Our waiter was inserted. Let's resolve the init closure.
-                match init() {
-                    Ok(value) => {
-                        *lock = Some(Ok(value.clone()));
-                        Initialized(value)
-                    }
-                    Err(e) => {
-                        let err: ErrorObject = Arc::new(e);
-                        *lock = Some(Err(Arc::clone(&err)));
-                        self.remove_waiter(&key, type_id);
-                        InitErr(err.downcast().unwrap())
+            match self.try_insert_waiter(key, type_id, &waiter) {
+                None => {
+                    // Our waiter was inserted. Let's resolve the init future.
+                    // Catching panic is safe here as we do not try to resolve the future again.
+                    match catch_unwind(AssertUnwindSafe(init)) {
+                        // Resolved.
+                        Ok(value) => return post_init(key, value, &mut lock),
+                        // Panicked.
+                        Err(payload) => {
+                            *lock = None;
+                            // Remove the waiter so that others can retry.
+                            self.remove_waiter(key, type_id);
+                            resume_unwind(payload);
+                        } // The lock will be unlock here.
                     }
                 }
-            }
-            Some(res) => {
-                // Somebody else's waiter already exists. Drop our write lock and wait
-                // for a read lock to become available.
-                std::mem::drop(lock);
-                match &*res.read() {
-                    Some(Ok(value)) => ReadExisting(value.clone()),
-                    Some(Err(e)) => InitErr(Arc::clone(e).downcast().unwrap()),
-                    None => unreachable!(),
+                Some(res) => {
+                    // Somebody else's waiter already exists. Drop our write lock and wait
+                    // for a read lock to become available.
+                    std::mem::drop(lock);
+                    match &*res.read() {
+                        Some(Ok(value)) => return ReadExisting(value.clone()),
+                        Some(Err(e)) => return InitErr(Arc::clone(e).downcast().unwrap()),
+                        // None means somebody else's init future has been panicked.
+                        // Retry from the beginning.
+                        None => continue,
+                    }
                 }
             }
         }
@@ -105,6 +135,7 @@ where
         self.waiters.remove(&(key, type_id));
     }
 
+    #[inline]
     fn try_insert_waiter(
         &self,
         key: &Arc<K>,

